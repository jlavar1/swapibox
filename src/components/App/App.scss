.grid-container {
  display: grid;
  grid-template-columns: 1fr 1fr 1fr 1fr;
  grid-template-rows: 1fr 0.4fr 1.6fr;
  grid-template-areas: 
    "movieCrawl Header    Header    Header" 
    "movieCrawl Buttons   Buttons   Buttons" 
    "movieCrawl Card-Area Card-Area Card-Area";
}

.movieCrawl {
  grid-area: movieCrawl;
}

.Header {
  grid-area: Header;
}

.Buttons {
  grid-area: Buttons;
}

.Card-Area {
  grid-area: Card-Area;
}