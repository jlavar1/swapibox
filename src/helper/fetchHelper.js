import React from 'react';

// const fetchHelper = (url) => {
//   fetch(url)
//     .then(response => response.json())
//     .then(data => data.opening_crawl)
//     .then(openCrawl => this.setState({openCrawl}))
//     .catch(error => this.setState({error: error.message}))
// }

const fetchHelper = async (url) => {
  const response = await fetch(url)
  try {
    return response.json()
  } catch {
    throw Error('Failed to retrieve data, please try again')
  }
}

export default fetchHelper;