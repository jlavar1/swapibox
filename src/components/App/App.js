import React, { Component } from 'react';
import Header from '../Header/Header';
import Button from '../Button/Button';
import CardContainer from '../CardContainer/CardContainer';
import Crawl from '../Crawl/Crawl';
import fetchHelper from '../../helper/fetchHelper';


class App extends Component {
  constructor() {
    super();
    this.state = {
      movieCrawl: {},
      error: '',
      vehicles: [],
      planets: [],
      people: [],
      enabled: ''
    }
  }

  componentDidMount() {
    this.fetchCrawl()
  }

  randomizeFilm = () => {
    return Math.floor(Math.random() * 8);
  }
  
  fetchCrawl = async () => {
    const url = `https://swapi.co/api/films/${this.randomizeFilm()}/`
    const data = await fetchHelper(url)
    this.setState({
      movieCrawl: {
        title: data.title,
        crawl: data.opening_crawl,
        release_date: data.release_date
      }
    })
  }

  fetchCategory = (category) => {
    switch(category) {
      case 'people':
        this.fetchPeople()
        break;
      case 'planets':
        this.fetchPlanets();
        break;
      case 'vehicles':
        this.fetchVehicles();
        break;
      default:
        // do nothing
    }
  }

  enableCategory = (category) => {
    this.setState({
      enabled: category
    })
  }

  fetchPeople = async () => {
    const url = 'https://swapi.co/api/people/';
    const people = await fetchHelper(url)
    const mappedPeople = await this.streamlinePeople(people.results)
    this.setState({
      people: mappedPeople
    })
  }

  streamlinePeople = (people) => {
    const streamlinedPeople = people.map(async (person) => {
      const fetchedHomeworld = await this.fetchHomeworld(person.homeworld)
      const fetchedSpecies = await this.fetchSpecies(person.species)
      return {
        name: person.name,
        homeworld: fetchedHomeworld.name,
        species: fetchedSpecies.classification,
        homeworld_population: fetchedHomeworld.population
      }
    })
    return Promise.all(streamlinedPeople)
  }

  fetchHomeworld = async (homeworlds) => {
    const fetchedHomeworld = await fetchHelper(homeworlds)
    const planet = await fetchedHomeworld
    return planet
  }

  fetchSpecies = async (species) => {
    const fetchedSpecies = await fetchHelper(species)
    const speciesData = await fetchedSpecies
    return speciesData
  }

  fetchVehicles = async () => {
    const url = 'https://swapi.co/api/vehicles/'
    const vehicles = await fetchHelper(url)
    const mapVehicles = await this.streamlineVehicles(vehicles.results);
    this.setState({
      vehicles: mapVehicles
    })
  }
  
  streamlineVehicles = vehicles => {
    const streamlinedVehicles = vehicles.map(vehicle => {
      return {
        name: vehicle.name,
        model: vehicle.model,
        class: vehicle.vehicle_class,
        numberOfPassengers: vehicle.passengers,
        id: vehicle.created
      }
    })
    return streamlinedVehicles
  }

  fetchPlanets = async () => {
    const url = 'https://swapi.co/api/planets/';
    const planetsPromised = await fetchHelper(url)
    const planets = await planetsPromised.results
    const mappedPlanets = await this.streamlinedPlanets(planets)
    this.setState({
      planets: mappedPlanets
    })
  }

  streamlinedPlanets = planets => {
    const streamlinedPlanets = planets.map( async (planet) => {
      const residents = await this.fetchResidents(planet.residents)
      return {
        name: planet.name,
        terrain: planet.terrain,
        population: planet.population,
        climate: planet.climate,
        residents: residents
      }
    })
    return Promise.all(streamlinedPlanets)
  }

  fetchResidents = (residents) => {
    let unresolvedResidents = residents.map(async (resident) => {
      const fetchedResident = await fetchHelper(resident)
      return fetchedResident.name
    })
    return Promise.all(unresolvedResidents)
  }

  render() {
    const { enabled, movieCrawl, error, vehicles, people, planets } = this.state
      return (
        <div className="grid-container">
          <div className="movieCrawl">
            <Crawl movieCrawl={movieCrawl}/>
          </div>
          <div className="Header">
            <Header/>
          </div>
          <div className="Buttons">
            <Button 
              fetchCategory={this.fetchCategory}
              enableCategory={this.enableCategory}
              vehicles={vehicles}
              people={people}
              planets={planets}/>
          </div>
          <div className="Card-Area">
            <CardContainer 
              vehicles={vehicles}
              people={people}
              planets={planets}
              enabled={enabled}
              category={this.state}/>
          </div>
          { error && error }
        </div>
      )
  }
}

export default App;
