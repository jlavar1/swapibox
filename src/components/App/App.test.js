import React from 'react';
import App from './App';
import { shallow } from 'enzyme';
import * as api from '../../helper/fetchHelper';

describe('App', () => {
  let wrapper;
  let mockData;

  beforeEach(() => {
    wrapper = shallow(
      <App />
    );

  });
  
  it('should match the snapshot', () => {
    expect(wrapper).toMatchSnapshot()
  });

  it('should have the proper default state', () => {
    expect(wrapper.state()).toEqual({
      movieCrawl: {},
      error: '',
      vehicles: [],
      planets: [],
      people: []
    });
  });

  it('should call fetchCrawl() when App mounts', () => { 
    wrapper.instance().fetchCrawl = jest.fn();

    wrapper.instance().componentDidMount();

    expect(wrapper.instance().fetchCrawl).toBeCalled();
  });

  it('should call fetchHelper in fetchCrawl', async () => {
    const url = 'https://swapi.co/';
    mockData = {
      title: 'The Force',
      opening_crawl: 'A long time ago...',
      release_date: 'March 29, 1991'
    }


    
    api.fetchHelper = jest.fn(() => Promise.resolve(mockData))

    await wrapper.instance().fetchCrawl()

    expect(api.fetchHelper).toBeCalled()
  });
})