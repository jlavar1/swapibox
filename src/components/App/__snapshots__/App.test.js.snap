// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App should match the snapshot 1`] = `
<div
  className="grid-container"
>
  <div
    className="movieCrawl"
  >
    <Crawl
      movieCrawl={Object {}}
    />
  </div>
  <div
    className="Header"
  >
    <Header />
  </div>
  <div
    className="Buttons"
  >
    <Button
      enableCategory={[Function]}
      fetchCategory={[Function]}
      people={Array []}
      planets={Array []}
      vehicles={Array []}
    />
  </div>
  <div
    className="Card-Area"
  >
    <CardContainer
      category={
        Object {
          "enabled": "",
          "error": "",
          "movieCrawl": Object {},
          "people": Array [],
          "planets": Array [],
          "vehicles": Array [],
        }
      }
      enabled=""
      people={Array []}
      planets={Array []}
      vehicles={Array []}
    />
  </div>
</div>
`;
